lib/rule-context.js

```diff
@@ -61,6 +62,73 @@ const PASSTHROUGHS = [
 //------------------------------------------------------------------------------
 
 /**
+ * Compares items in a fixes array by range.
+ * @param {Fix} a The first message.
+ * @param {Fix} b The second message.
+ * @returns {int} -1 if a comes before b, 1 if a comes after b, 0 if equal.
+ * @private
+ */
+function compareFixesByRange(a, b) {
+    return a.range[0] - b.range[0] || a.range[1] - b.range[1];
+}
+
+/**
+ * Merges the given fixes array into one.
+ * @param {Fix[]} fixes The fixes to merge.
+ * @param {SourceCode} sourceCode The source code object to get the text between fixes.
+ * @returns {void}
+ */
+function mergeFixes(fixes, sourceCode) {
+    if (fixes.length === 0) {
+        return null;
+    }
+    if (fixes.length === 1) {
+        return fixes[0];
+    }
+
+    fixes.sort(compareFixesByRange);
+
+    const originalText = sourceCode.text;
+    const start = fixes[0].range[0];
+    const end = fixes[fixes.length - 1].range[1];
+    let text = "";
+    let lastPos = start;
+
+    for (const fix of fixes) {
+        assert(fix.range[0] >= lastPos, "Fix objects must not be overlapped in a report.");
+
+        text += originalText.slice(lastPos, fix.range[0]);
```

Comment:
Some fix ranges, like [`unicode-bom`](https://github.com/eslint/eslint/blob/a8e1c1ccf2a212612f8aa63d7463dfa9acd920b6/lib/rules/unicode-bom.js#L57), use `-1` as the start to indicate that the fix should replace the BOM. I think that will cause a bug here because `.slice(-1)` will take the *last* character in the string.